crates/sui-types/src/zk_login_authenticator.rs

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) 2021, Facebook, Inc. and its affiliates
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::{
+    base_types::SuiAddress,
+    crypto::{Signature, SignatureScheme, SuiSignature},
+    error::SuiError,
+    signature::{AuthenticatorTrait, AuxVerifyData},
+    zk_login_util::{find_jwk_by_kid, AddressParams, DEFAULT_WHITELIST},
+};
+use fastcrypto::rsa::Encoding as OtherEncoding;
+use fastcrypto::rsa::RSAPublicKey;
+use fastcrypto::rsa::RSASignature;
+use fastcrypto::{error::FastCryptoError, rsa::Base64UrlUnpadded, traits::ToFromBytes};
+use fastcrypto_zkp::bn254::zk_login::verify_zk_login_proof_with_fixed_vk;
+use fastcrypto_zkp::bn254::zk_login::{is_claim_supported, AuxInputs, PublicInputs, ZkLoginProof};
+use once_cell::sync::OnceCell;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use shared_crypto::intent::IntentMessage;
+use std::hash::Hash;
+use std::hash::Hasher;
+
+#[cfg(test)]
+#[cfg(feature = "test-utils")]
+#[path = "unit_tests/zk_login_authenticator_test.rs"]
+mod zk_login_authenticator_test;
+
+/// An zk login authenticator with all the necessary fields.
+#[derive(Debug, Clone, JsonSchema, Serialize, Deserialize)]
+pub struct ZkLoginAuthenticator {
+    proof: ZkLoginProof,
+    public_inputs: PublicInputs,
+    aux_inputs: AuxInputs,
+    user_signature: Signature,
+    #[serde(skip)]
+    pub bytes: OnceCell<Vec<u8>>,
+}
+
+impl ZkLoginAuthenticator {
+    /// Create a new [struct ZkLoginAuthenticator] with necessary fields.
+    pub fn new(
+        proof: ZkLoginProof,
+        public_inputs: PublicInputs,
+        aux_inputs: AuxInputs,
+        user_signature: Signature,
+    ) -> Self {
+        Self {
+            proof,
+            public_inputs,
+            aux_inputs,
+            user_signature,
+            bytes: OnceCell::new(),
+        }
+    }
+
+    pub fn get_address_seed(&self) -> &str {
+        self.aux_inputs.get_address_seed()
+    }
+
+    pub fn get_address_params(&self) -> AddressParams {
+        AddressParams::new(
+            self.aux_inputs.get_iss().to_string(),
+            self.aux_inputs.get_claim_name().to_string(),
+        )
+    }
+}
+
+/// Necessary trait for [struct SenderSignedData].
+impl PartialEq for ZkLoginAuthenticator {
+    fn eq(&self, other: &Self) -> bool {
+        self.as_ref() == other.as_ref()
+    }
+}
+
+/// Necessary trait for [struct SenderSignedData].
+impl Eq for ZkLoginAuthenticator {}
```

Comment:
should we ignore bytes (if set or not), or only compare set bytes?

Reply:
yes this is just comparing hte bytes (i believe eq is derived from partial eq)